source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.h

```diff
@@ -0,0 +1,219 @@
+
+#include "absl/status/status.h"
+
+namespace Envoy {
+namespace Upstream {
+
+using ClientSideWeightedRoundRobinLbProto = envoy::extensions::load_balancing_policies::
+    client_side_weighted_round_robin::v3::ClientSideWeightedRoundRobin;
+using OrcaLoadReportProto = xds::data::orca::v3::OrcaLoadReport;
+
+/**
+ * Load balancer config used to wrap the config proto.
+ */
+class ClientSideWeightedRoundRobinLbConfig : public Upstream::LoadBalancerConfig {
+public:
+  ClientSideWeightedRoundRobinLbConfig(const ClientSideWeightedRoundRobinLbProto& lb_proto,
+                                       Event::Dispatcher& main_thread_dispatcher);
+
+  // Parameters for weight calculation from Orca Load report.
+  std::vector<std::string> metric_names_for_computing_utilization;
+  double error_utilization_penalty;
+  // Timing parameters for the weight update.
+  std::chrono::milliseconds blackout_period;
+  std::chrono::milliseconds weight_expiration_period;
+  std::chrono::milliseconds weight_update_period;
+
+  Event::Dispatcher& main_thread_dispatcher_;
+};
+
+/**
+ * A client side weighted round robin load balancer. When in weighted mode, EDF
+ * scheduling is used. When in not weighted mode, simple RR index selection is
+ * used.
+ */
+class ClientSideWeightedRoundRobinLoadBalancer : public Upstream::ThreadAwareLoadBalancer,
+                                                 protected Logger::Loggable<Logger::Id::upstream> {
+public:
+  // This struct is used to store the client side data for the host. Hosts are
+  // not shared between different clusters, but are shared between load
+  // balancer instances on different threads.
+  struct ClientSideHostLbPolicyData : public Envoy::Upstream::Host::HostLbPolicyData {
+    ClientSideHostLbPolicyData() = default;
+    ClientSideHostLbPolicyData(uint32_t weight, MonotonicTime non_empty_since,
+                               MonotonicTime last_update_time)
+        : weight_(weight), non_empty_since_(non_empty_since), last_update_time_(last_update_time) {}
+    virtual ~ClientSideHostLbPolicyData() = default;
+    // Update the weight and timestamps for first and last update time.
+    void updateWeightNow(uint32_t weight, const MonotonicTime& now) {
+      absl::MutexLock lock(&mu_);
+      weight_ = weight;
+      last_update_time_ = now;
+      if (non_empty_since_ == kDefaultNonEmptySince) {
+        non_empty_since_ = now;
+      }
+    }
+    absl::Mutex mu_;
+    // Weight as calculated from the last load report.
+    uint32_t weight_ ABSL_GUARDED_BY(mu_) = 1;
+    // Time when the weight is first updated. The weight is invalid if it is within of
+    // `blackout_period_`.
+    MonotonicTime non_empty_since_ ABSL_GUARDED_BY(&mu_) = kDefaultNonEmptySince;
+    // Time when the weight is last updated. The weight is invalid if it is outside of
+    // `expiration_period_`.
+    MonotonicTime last_update_time_ ABSL_GUARDED_BY(&mu_) = kDefaultLastUpdateTime;
```

Comment:
Hmmm, will atomic provides better performance? Both time point and the weight are simple integer which is pretty good to use atomic. I think if atomic is used, then no lock is introduced in the whole feature?

Reply:
Sounds good. Done.